File: frontend/lib/server/agent/responses-runtime.ts
// frontend/lib/server/agent/responses-runtime.ts
// Responses-shaped 流循环：逐字节忠实移植自 Python src/kodeks/responses_runtime.py。
// runResponsesToolLoop（N 轮驱动器）/ liveResponsesEvents（组合 M1 桥 + M0 config）/ error helpers / pseudo 检测。
//
// 保真红线（见 30-runtime-loop.md「工具循环驱动逻辑」、80-oracle §B/§D）：
//  · 每轮重建 input = buildResponsesInputFromTranscript；maxTurns=12。
//  · 流消费分支：output_text.delta→text_delta(非空)；output_item.done→handleOutputItem；
//    response.failed→error(moonbridge_upstream_failed) return；error→error return；
//    response.completed→若 toolMessages/waiting/halt 则 break，否则 pseudo 检测→complete_assistant_turn 收尾 return。
//  · 轮末决策：halt→return；waiting→return；toolMessages→append+continue；!completed→error(stream ended) return；else return。
//  · 耗尽→error("Model tool loop exceeded the maximum turn limit.")。
//  · response_id 回退 "resp_python"；pseudo 检测含 "<tool_call" 或 'type="tool_calls"'（小写）。
import {
  ModelConfigurationError,
  loadModelRuntimeEnv,
  readChatCompletionsApiKey,
  readChatCompletionsBaseUrl,
  readChatCompletionsConfig,
  resolveModelClientOptions,
  type RuntimeEnv,
} from '../config'
import {
  fetchChatCompletionsStream,
  fromDeepseekStream,
  toDeepseekChatRequest,
} from '../bridge'
import { defaultToolDefinitions } from '../tools/schemas'
import type { KodeksDatabase } from '../storage'
import {
  buildResponsesInputFromTranscript,
  type ResponsesInputItem,
} from './conversation-state'
import {
  appendToolContinuationMessages,
  handleOutputItem,
  ToolRoundState,
  type RuntimeEvent,
  type ToolRegistryLike,
} from './tool-loop'

/** 注入点产出的 Responses 事件流：同步可迭代或异步可迭代（移植 ResponsesEventStream，responses_runtime.py:30）。 */
export type ResponsesEventStream =
  | Iterable<Record<string, unknown>>
  | AsyncIterable<Record<string, unknown>>

/** 假模型/真实模型注入工厂：(body, env) → 事件流（移植 ResponsesEventFactory，responses_runtime.py:31-33）。 */
export type ResponsesEventFactory = (
  body: Record<string, unknown>,
  env: RuntimeEnv,
) => ResponsesEventStream

/** 收尾工厂：(assistantText, responseId) → 事件异步迭代（移植 CompletionEventFactory，responses_runtime.py:34）。 */
export type CompletionEventFactory = (
  assistantText: string,
  responseId: string,
) => AsyncIterable<RuntimeEvent>

/** runResponsesToolLoop 入参（命名参数，对应 Python keyword-only）。 */
export interface RunResponsesToolLoopArgs {
  body: Record<string, unknown>
  runtimeBody: Record<string, unknown>
  database: KodeksDatabase
  workspaceRoot: string
  runtimeEnv: RuntimeEnv
  sessionId: string
  registry: ToolRegistryLike
  completeAssistantTurn: CompletionEventFactory
  responsesEventFactory: ResponsesEventFactory | null
  maxToolLoopTurns: number
}

/**
 * 把 Responses-shaped 模型事件经本地工具 continuation 跑完（移植 run_responses_tool_loop，responses_runtime.py:37-151）。
 * 每轮重建 input → 调注入工厂或 liveResponsesEvents → 消费事件 → 决定继续/暂停/终止。
 */
export async function* runResponsesToolLoop(
  args: RunResponsesToolLoopArgs,
): AsyncGenerator<RuntimeEvent> {
  const {
    runtimeBody,
    database,
    workspaceRoot,
    runtimeEnv,
    sessionId,
    registry,
    completeAssistantTurn,
    responsesEventFactory,
    maxToolLoopTurns,
  } = args

  for (let turnIndex = 0; turnIndex < maxToolLoopTurns; turnIndex += 1) {
    runtimeBody.input = (await buildResponsesInputFromTranscript(
      database,
      sessionId,
    )) as ResponsesInputItem[]
    const responsesEvents =
      responsesEventFactory !== null
        ? responsesEventFactory(runtimeBody, runtimeEnv)
        : liveResponsesEvents(runtimeBody, runtimeEnv)
    let assistantText = ''
    let completed = false
    const toolState = new ToolRoundState()
    for await (const event of aiter(responsesEvents)) {
      const eventType = event.type
      if (eventType === 'response.output_text.delta') {
        const delta = String(event.delta || '')
        if (delta) {
          assistantText += delta
          yield {
            type: 'text_delta',
            delta,
            session_id: sessionId,
          }
        }
        continue
      }

      if (eventType === 'response.output_item.done') {
        for await (const toolEvent of handleOutputItem(
          event.item,
          registry,
          database,
          workspaceRoot,
          runtimeEnv,
          sessionId,
          toolState,
        )) {
          yield toolEvent
        }
        continue
      }

      if (eventType === 'response.failed') {
        yield errorEvent(responseErrorMessage(event), sessionId, 'moonbridge_upstream_failed')
        return
      }

      if (eventType === 'error') {
        yield errorEvent(streamErrorMessage(event), sessionId)
        return
      }

      if (eventType === 'response.completed') {
        completed = true
        if (
          toolState.toolMessages.length > 0 ||
          toolState.waitingForApproval ||
          toolState.haltToolLoop
        ) {
          break
        }
        if (looksLikePseudoToolCall(assistantText)) {
          yield errorEvent(
            'Model returned tool-call text instead of a native function_call event.',
            sessionId,
            'model_returned_pseudo_tool_call',
          )
          return
        }
        const response = event.response
        const responseId =
          isDict(response) && response.id !== undefined && response.id !== null
            ? String(response.id)
            : 'resp_python'
        for await (const completionEvent of completeAssistantTurn(assistantText, responseId)) {
          yield completionEvent
        }
        return
      }
    }

    if (toolState.haltToolLoop) {
      return
    }
    if (toolState.waitingForApproval) {
      return
    }
    if (toolState.toolMessages.length > 0) {
      await appendToolContinuationMessages(
        database,
        sessionId,
        assistantText,
        toolState.reasoningContent,
        toolState.toolCalls,
        toolState.toolMessages,
      )
      continue
    }
    if (!completed) {
      yield errorEvent('Model stream ended before completion.', sessionId)
      return
    }
    return
  }

  yield errorEvent('Model tool loop exceeded the maximum turn limit.', sessionId)
}

/**
 * 从已配置的模型路由生成真实 Responses-shaped 事件（移植 live_responses_events，responses_runtime.py:154-193）。
 * 仅 moonbridge provider 才继续；缺 provider/配置/api_key 抛 ModelConfigurationError。
 * 上游 request 形状（键序固定）经 toDeepseekChatRequest → fetchChatCompletionsStream → fromDeepseekStream。
 */
export async function* liveResponsesEvents(
  body: Record<string, unknown>,
  runtimeEnv: RuntimeEnv,
): AsyncGenerator<Record<string, unknown>> {
  // 解析模型 env（合并配置文件/.env，模型 ref 决定 DeepSeek 模型）。
  const modelEnv = loadModelRuntimeEnv(runtimeEnv, body.model)
  const modelOptions = resolveModelClientOptions(
    modelEnv,
    body.reasoning_effort,
    body.provider,
  )
  if (modelOptions === null) {
    throw new ModelConfigurationError(
      'An OpenAI-compatible Chat Completions provider is required. Set API_KEY or DEEPSEEK_API_KEY for the MoonBridge route.',
    )
  }
  if (modelOptions.provider !== 'moonbridge') {
    throw new ModelConfigurationError('Unsupported model provider.')
  }

  const upstream = readChatCompletionsConfig(modelEnv)
  const missing = upstream.missing
  if (Array.isArray(missing) && missing.length > 0) {
    throw new ModelConfigurationError(
      `Missing upstream Chat Completions configuration: ${missing.join(', ')}.`,
    )
  }
  const apiKey = readChatCompletionsApiKey(modelEnv)
  if (apiKey === undefined) {
    throw new ModelConfigurationError('API_KEY or DEEPSEEK_API_KEY is required.')
  }

  // request 键序逐字对齐 Python（responses_runtime.py:180-187）。
  const request: Record<string, unknown> = {
    model: body.model || modelOptions.model,
    input: body.input || '',
    instructions: body.instructions || '',
    tools: defaultToolDefinitions(body.mode === 'plan'),
    reasoning: { effort: body.reasoning_effort || 'high' },
    stream: true,
  }
  const payload = toDeepseekChatRequest(request, String(upstream.model))
  // TS 桥的 fetch 签名为 (payload, {apiKey, baseURL})；baseURL 由 config 解析（与 Python 传 model_env 等价）。
  const baseURL = readChatCompletionsBaseUrl(modelEnv)
  const stream = fetchChatCompletionsStream(payload, { apiKey, baseURL })
  for await (const event of fromDeepseekStream(stream, { model: String(request.model) })) {
    yield event
  }
}

/** 从失败的 Responses 事件读取用户可见信息（移植 _response_error_message，responses_runtime.py:196-204）。 */
function responseErrorMessage(event: Record<string, unknown>): string {
  const response = event.response
  if (isDict(response)) {
    const error = response.error
    if (isDict(error) && typeof error.message === 'string') {
      return String(error.message)
    }
  }
  return 'Model stream failed.'
}

/** 从 Responses error 流事件读取用户可见信息（移植 _stream_error_message，responses_runtime.py:207-211）。 */
function streamErrorMessage(event: Record<string, unknown>): string {
  const message = event.message
  return typeof message === 'string' && message ? String(message) : 'Model stream failed.'
}

/** 判断可见文本是否含伪造的序列化工具调用（移植 _looks_like_pseudo_tool_call，responses_runtime.py:214-218）。 */
function looksLikePseudoToolCall(text: string): boolean {
  const lowered = text.toLowerCase()
  return lowered.includes('<tool_call') || lowered.includes('type="tool_calls"')
}

/** 构建一个运行时 error 事件（移植 _error_event，responses_runtime.py:221-231）。 */
function errorEvent(message: string, sessionId: string, code = 'runtime_error'): RuntimeEvent {
  return {
    type: 'error',
    message,
    code,
    session_id: sessionId,
  }
}

/** 遍历同步或异步 Responses 事件流（移植 _aiter，responses_runtime.py:234-242）。 */
async function* aiter(stream: ResponsesEventStream): AsyncGenerator<Record<string, unknown>> {
  if (isSyncIterable(stream)) {
    for (const item of stream) {
      yield item
    }
    return
  }
  for await (const item of stream) {
    yield item
  }
}

/** 判断一个值是否同步可迭代（有 Symbol.iterator）。 */
function isSyncIterable(value: ResponsesEventStream): value is Iterable<Record<string, unknown>> {
  return typeof (value as Iterable<Record<string, unknown>>)[Symbol.iterator] === 'function'
}

/** 复刻 Python `isinstance(x, dict)`：普通对象（非 null、非数组）。 */
function isDict(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

File: src/kodeks/responses_runtime.py
"""Responses-shaped stream loop for Kodeks chat turns."""

from __future__ import annotations

from collections.abc import AsyncIterator, Callable, Iterable, Mapping
from typing import Any, cast

from .config import (
    ModelConfigurationError,
    load_model_runtime_env,
    read_chat_completions_api_key,
    read_chat_completions_config,
    resolve_model_client_options,
)
from .conversation_state import build_responses_input_from_transcript
from .providers.bridge import (
    fetch_chat_completions_stream,
    from_deepseek_stream,
    to_deepseek_chat_request,
)
from .responses_tool_loop import (
    ToolRegistryLike,
    ToolRoundState,
    append_tool_continuation_messages,
    handle_output_item,
)
from .storage import KodeksDatabase
from .tools.schemas import default_tool_definitions

ResponsesEventStream = AsyncIterator[dict[str, Any]] | Iterable[dict[str, Any]]
ResponsesEventFactory = Callable[
    [Mapping[str, Any], Mapping[str, str | None]], ResponsesEventStream
]
CompletionEventFactory = Callable[[str, str], AsyncIterator[dict[str, Any]]]


async def run_responses_tool_loop(
    *,
    body: Mapping[str, Any],
    runtime_body: dict[str, Any],
    database: KodeksDatabase,
    workspace_root: str,
    runtime_env: Mapping[str, str | None],
    session_id: str,
    registry: ToolRegistryLike,
    complete_assistant_turn: CompletionEventFactory,
    responses_event_factory: ResponsesEventFactory | None,
    max_tool_loop_turns: int,
) -> AsyncIterator[dict[str, Any]]:
    """Run Responses-shaped model events through local tool continuation."""

    for _turn_index in range(max_tool_loop_turns):
        runtime_body["input"] = build_responses_input_from_transcript(
            database, session_id
        )
        responses_events = (
            responses_event_factory(runtime_body, runtime_env)
            if responses_event_factory is not None
            else live_responses_events(runtime_body, runtime_env)
        )
        assistant_text = ""
        completed = False
        tool_state = ToolRoundState()
        async for event in _aiter(responses_events):
            event_type = event.get("type")
            if event_type == "response.output_text.delta":
                delta = str(event.get("delta") or "")
                if delta:
                    assistant_text += delta
                    yield {
                        "type": "text_delta",
                        "delta": delta,
                        "session_id": session_id,
                    }
                continue

            if event_type == "response.output_item.done":
                async for tool_event in handle_output_item(
                    event.get("item"),
                    registry,
                    database,
                    workspace_root,
                    runtime_env,
                    session_id,
                    tool_state,
                ):
                    yield tool_event
                continue

            if event_type == "response.failed":
                yield _error_event(
                    _response_error_message(event),
                    session_id,
                    "moonbridge_upstream_failed",
                )
                return

            if event_type == "error":
                yield _error_event(_stream_error_message(event), session_id)
                return

            if event_type == "response.completed":
                completed = True
                if (
                    tool_state.tool_messages
                    or tool_state.waiting_for_approval
                    or tool_state.halt_tool_loop
                ):
                    break
                if _looks_like_pseudo_tool_call(assistant_text):
                    yield _error_event(
                        (
                            "Model returned tool-call text instead of a native "
                            "function_call event."
                        ),
                        session_id,
                        "model_returned_pseudo_tool_call",
                    )
                    return
                response = event.get("response")
                response_id = (
                    str(response.get("id"))
                    if isinstance(response, dict) and response.get("id") is not None
                    else "resp_python"
                )
                async for completion_event in complete_assistant_turn(
                    assistant_text, response_id
                ):
                    yield completion_event
                return

        if tool_state.halt_tool_loop:
            return
        if tool_state.waiting_for_approval:
            return
        if tool_state.tool_messages:
            append_tool_continuation_messages(
                database=database,
                session_id=session_id,
                assistant_text=assistant_text,
                reasoning_content=tool_state.reasoning_content,
                tool_calls=tool_state.tool_calls,
                tool_messages=tool_state.tool_messages,
            )
            continue
        if not completed:
            yield _error_event("Model stream ended before completion.", session_id)
            return
        return

    yield _error_event("Model tool loop exceeded the maximum turn limit.", session_id)


async def live_responses_events(
    body: Mapping[str, Any], runtime_env: Mapping[str, str | None]
) -> AsyncIterator[dict[str, Any]]:
    """Create live Responses-shaped events from configured model routing."""

    model_env = load_model_runtime_env(runtime_env, body.get("model"))
    model_options = resolve_model_client_options(
        model_env, body.get("reasoning_effort"), body.get("provider")
    )
    if model_options is None:
        raise ModelConfigurationError(
            "An OpenAI-compatible Chat Completions provider is required. Set API_KEY or DEEPSEEK_API_KEY for the MoonBridge route."
        )
    if model_options["provider"] != "moonbridge":
        raise ModelConfigurationError("Unsupported model provider.")

    upstream = read_chat_completions_config(model_env)
    if upstream["missing"]:
        missing = cast(list[str], upstream["missing"])
        raise ModelConfigurationError(
            f"Missing upstream Chat Completions configuration: {', '.join(missing)}."
        )
    api_key = read_chat_completions_api_key(model_env)
    if api_key is None:
        raise ModelConfigurationError("API_KEY or DEEPSEEK_API_KEY is required.")

    request = {
        "model": body.get("model") or model_options["model"],
        "input": body.get("input") or "",
        "instructions": body.get("instructions") or "",
        "tools": default_tool_definitions(body.get("mode") == "plan"),
        "reasoning": {"effort": body.get("reasoning_effort") or "high"},
        "stream": True,
    }
    payload = to_deepseek_chat_request(request, str(upstream["model"]))
    async for event in from_deepseek_stream(
        fetch_chat_completions_stream(payload, api_key, model_env),
        model=str(request["model"]),
    ):
        yield event


def _response_error_message(event: Mapping[str, Any]) -> str:
    """Read the user-visible message from a failed Responses event."""

    response = event.get("response")
    if isinstance(response, dict):
        error = response.get("error")
        if isinstance(error, dict) and isinstance(error.get("message"), str):
            return str(error["message"])
    return "Model stream failed."


def _stream_error_message(event: Mapping[str, Any]) -> str:
    """Read the user-visible message from a Responses error stream event."""

    message = event.get("message")
    return str(message) if isinstance(message, str) and message else "Model stream failed."


def _looks_like_pseudo_tool_call(text: str) -> bool:
    """Return whether visible text contains a fake serialized tool call."""

    lowered = text.lower()
    return "<tool_call" in lowered or 'type="tool_calls"' in lowered


def _error_event(
    message: str, session_id: str, code: str = "runtime_error"
) -> dict[str, Any]:
    """Build one runtime error event."""

    return {
        "type": "error",
        "message": message,
        "code": code,
        "session_id": session_id,
    }


async def _aiter(stream: ResponsesEventStream) -> AsyncIterator[dict[str, Any]]:
    """Iterate over sync or async Responses event streams."""

    if isinstance(stream, Iterable):
        for item in stream:
            yield item
        return
    async for item in stream:
        yield item
